zilliqa/src/consensus.rs

```diff
@@ -1081,16 +1061,14 @@ impl Consensus {
     }
 
     /// Check the validity of a block
-    fn check_block(&mut self, block: &Block) -> Result<()> {
+    fn check_block(&self, block: &Block) -> Result<()> {
         block.verify_hash()?;
 
         if block.view() == 0 {
             return Ok(());
         }
 
-        let Some(finalized_block) = self.get_block_by_view(self.finalized_view)? else {
-            return Err(MissingBlockError::from(self.finalized_view).into());
-        };
+        let Some(finalized_block) = self.get_block_by_view_locally(self.finalized_view)? else { return Err(MissingBlockError::from(self.finalized_view).into()); };
```

Comment:
What I mean for the final point is that you have implemented a manual way to wait for network events. You will have to do this for every struct that wants to be able to do it, but if we had a networking framework that returned promises the logic for waiting wouldn't have to belong in block store

Reply:
Ah I see what you mean. We don't really have a network abstraction - just message passing. Generally it wouldn't even make sense for all events to return a future, because not all broadcast events require a response.

I think requesting blocks and potentially transactions are really the only ones that need this right now, so I don't have any qualms with keeping it in the BlockStore - though if we need to generalise it in the future then I think the logic could be simply extracted into a generic NetworkRequest type struct, which basically does the same thing I've added to BlockStore now except with generic messages and data. (And the BlockStore would just hook into that, in that case.) Like I said though I'd leave that until we need it.